### DISCUSSION TITLE
Why are `remove` and `private` attributes mutually exclusive?

### DISCUSSION THREAD

**Discussion by Mixaz017** [twpayne/chezmoi#3415] [Q&A] [answered]
I manage a lot of files under `.config/` that are mainly used by Linux machines, but some of them are also used by Windows machines, so I can't ignore the entire `.config` just by checking the OS and must ignore all files/directories under `.config/` individually. I also use `lookPath` function to check if the corresponding program is installed, to avoid polluting the config directory with unused files.

The problem is, the `.config/` folder could be empty (especially on Windows) if none of the programs that use `.config/` directory is installed. However, I can't use `remove` attribute on it because `private` is already used, and they can't be used together.

So, why are they mutually exclusive in the first place? I feel like it's a valid use case.

Technically, I COULD check if none of the programs that use `.config/` is installed and ignore the entire directory, but that's a lot of work for something that could have been done if those attributes are compatible.

Alternative ways to solve my problem would be a proper reverse ignore (include/allowlist) support like #2852, but it seems like people are against that idea. Or some kind of "profiles" support like [dotdrop](https://github.com/deadc0de6/dotdrop) has would be cool, but that's too different from what chezmoi is doing.

**Comment by twpayne** [comment, marked as answer]
For directories, `remove_` declares that the directory should not exist. `private_` declares that the directory should exist and its permissions should only be readable by the user. Since these declarations disagree on whether the directory should exist or not, they are mutually exclusive.

The purpose of the `remove_` attribute is to make it easier to remove directory entries across multiple devices. Once the directory entry is removed from all of your devices you can remove the `remove_` entry from your source state. You can also use a [`.chezmoiremove` file](https://www.chezmoi.io/reference/special-files-and-directories/chezmoiremove/).

If you want to avoid creating irrelevant files in the first place then use `.chezmoiignore`. For example, if you only want `~/.config/git/config` on machines where git is installed you can put the following in your `.chezmoiignore` file:

```
{{ if not (lookPath "git") }}
.config/git
.config/git/**
{{ end }}
```

I would also ask: does it really matter if  `~/.config` exists but is empty?

**Reply by ernieIzde8ski** [reply]
good answer, but you need parentheses around that function, otherwise the templating engine gets mad at you:

```
{{ if not (lookPath "git") }}
```

**Reply by twpayne** [reply]
Thanks! I've edited my comment to include the fix.

**Comment by Mixaz017** [comment]
> The purpose of the `remove_` attribute is to make it easier to remove directory entries across multiple devices.

Thank you for clarifying the purpose of the attribute. It seems that my use of the attribute is unintended/misuse.

> I would also ask: does it really matter if `~/.config` exists but is empty?

To me, yes. If a file/directory is not used, it shouldn't exist.
I forgot to mention that I have a few other cases of this problem in other locations, and they make it harder to find the actual files/directories I'm looking for because unused files clutter path completions, `ls` outputs and file managers.

I still think the functionality to ignore empty destination directories should exist (or even be the default behavior). Something like an attribute that is functionally similar to `remove` but does not actively remove the destination directory would be very useful.

**Reply by Lockszmith-GH** [reply]
You can add an `run_after_` script that checks if a folder is empty and remove it. But the easiest way would probably be a templated `.chezmoiremove` file as **twpayne** mentioned.

**Reply by Mixaz017** [reply]
Thank you for your solutions. The problem with those is that it feels like a workaround, not a proper solution. You need to use different commands for different OS with `run` scripts which means it's another OS if else check, and it's unintuitive. Also, as twpayne mentioned, `.chezmoiremove`d files should not exist, so it's semantically different from what I want to do. Both solutions also mess up existing directories that weren't originally managed by chezmoi.

**Reply by Lockszmith-GH** [reply]
chezmoi, unlike git, actually cares about folders, even if they are empty - plus, you are talking about the destination/target, not the source repo.

If I'm understanding - you are looking for a new flag, let's call it `nonempty`, which will act as remove-if-empty logic.
Don't know if this runs against the grain with tom, as he's the master of this domain, but sounds doable.

There is some logic that needs to be thought off though.
- If a file or folder exist within the target, the directory must remain.
- the directory structure will need to be evaluated depth first, as clearing a subdir needs to happen first.

Is this what you thought about?

**Reply by twpayne** [reply]
Thank you both, I agree with both of you. I think https://github.com/twpayne/chezmoi/pull/3420 fixes it. Would you be able to test this?

**Reply by Mixaz017** [reply]
Thank you for taking your time to implement this. It worked perfectly in my testing.
After realizing technical problems @Lockszmith-GH pointed out, I think the amount of work needed to implement what I originally proposed is too much for a somewhat minor feature, so I consider #3420 to be the ideal solution here.
